Add ability to get boundaries (for words and uncategorized words)

conniey opened the pull request:
The existing `BreakIterator.Split` method returns strings. This is a problem when you want use the returned string token within a larger piece of text because Windows and icu have different collation implementations.  This results in weird behaviour such as [dotnet/corefx#9202](https://github.com/dotnet/corefx/issues/9202).

This also adds the option to return non-categorized word boundaries such as punctuation and spaces.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/sillsdev/icu-dotnet/9)
<!-- Reviewable:end -->

ermshiperete commented:
Reviewed 4 of 4 files at r1.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[source/icu.net/Boundary.cs, line 37 \[r1\]](https://reviewable.io:443/reviews/sillsdev/icu-dotnet/9#-KJnyceMCZ3N6ejNEK8k:-KJnyceMCZ3N6ejNEK8l:195682456) ([raw file](https://github.com/sillsdev/icu-dotnet/blob/4cd1a8d8485f767eaff2c512e739d44e6d46b918/source/icu.net/Boundary.cs#L37)):*
> ```Smalltalk
> 		public override bool Equals(object obj)
> 		{
> 			if (!(obj is Boundary))
> ```

You could use `as` and a check for `null` instead of those 6 lines, but it doesn't really matter:

    var compared = obj as Boundary;
    if (compared == null)
        return false;

---

*[source/icu.net/BreakIterator.cs, line 112 \[r1\]](https://reviewable.io:443/reviews/sillsdev/icu-dotnet/9#-KJo-18pF6M8OQI0HWtT:-KJo-18pF6M8OQI0HWtU:378220027) ([raw file](https://github.com/sillsdev/icu-dotnet/blob/4cd1a8d8485f767eaff2c512e739d44e6d46b918/source/icu.net/BreakIterator.cs#L112)):*
> ```Smalltalk
> 		/// <summary>
> 		/// Gets the sentence/line/word/character boundaries for the text. Spaces and punctuations
> 		/// are not returned for UBreakIteratorType.WORD.
> ```

The comment doesn't seem to match the implementation: this overload will never return spaces and punctuations regardless of the passed in `UBreakIteratorType`.

Later: oh, I see. It's the behaviour of ICU that spaces and punctuation only matter for `UBreakIteratorType.WORD`.  Maybe that should be added for clarification as a comment to the `includeSpacesAndPunctuation` parameter of the next overload.

---

*[source/icu.net.tests/BreakIteratorTests.cs, line 98 \[r1\]](https://reviewable.io:443/reviews/sillsdev/icu-dotnet/9#-KJo13oXDxeOAd3gH6PQ:-KJo13oXDxeOAd3gH6PR:-132095553) ([raw file](https://github.com/sillsdev/icu-dotnet/blob/4cd1a8d8485f767eaff2c512e739d44e6d46b918/source/icu.net.tests/BreakIteratorTests.cs#L98)):*
> ```Smalltalk
> 
> 		[Test]
> 		public void GetWordBoundaries()
> ```

These tests might be clearer to understand if you'd split them in two tests: something like `GetWordBoundaries_IncludeSpacesAndPunctuation` and `GetWordBoundaries_IgnoreSpacesAndPunctuation`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/sillsdev/icu-dotnet/9)*
<!-- Sent from Reviewable.io -->

conniey commented:
Review status: 1 of 4 files reviewed at latest revision, 3 unresolved discussions.

---

*[source/icu.net/Boundary.cs, line 37 \[r1\]](https://reviewable.io:443/reviews/sillsdev/icu-dotnet/9#-KJnyceMCZ3N6ejNEK8k:-KJqRfpK9sttBBtneyl9:-1411660648) ([raw file](https://github.com/sillsdev/icu-dotnet/blob/4cd1a8d8485f767eaff2c512e739d44e6d46b918/source/icu.net/Boundary.cs#L37)):*
<details><summary><i>Previously, ermshiperete (Eberhard Beilharz) wrote…</i></summary>
> You could use `as` and a check for `null` instead of those 6 lines, but it doesn't really matter:
> 
>     var compared = obj as Boundary;
>     if (compared == null)
>         return false;
</details>
It was non-nullable because it was a struct. I've changed it to a class though and fixed that method.

---

*[source/icu.net/BreakIterator.cs, line 112 \[r1\]](https://reviewable.io:443/reviews/sillsdev/icu-dotnet/9#-KJo-18pF6M8OQI0HWtT:-KJqS7t8I0-t4QWxldre:-499146009) ([raw file](https://github.com/sillsdev/icu-dotnet/blob/4cd1a8d8485f767eaff2c512e739d44e6d46b918/source/icu.net/BreakIterator.cs#L112)):*
<details><summary><i>Previously, ermshiperete (Eberhard Beilharz) wrote…</i></summary>
> The comment doesn't seem to match the implementation: this overload will never return spaces and punctuations regardless of the passed in `UBreakIteratorType`.
> 
> Later: oh, I see. It's the behaviour of ICU that spaces and punctuation only matter for `UBreakIteratorType.WORD`.  Maybe that should be added for clarification as a comment to the `includeSpacesAndPunctuation` parameter of the next overload.
</details>
Done.

---

*[source/icu.net.tests/BreakIteratorTests.cs, line 98 \[r1\]](https://reviewable.io:443/reviews/sillsdev/icu-dotnet/9#-KJo13oXDxeOAd3gH6PQ:-KJqS5uj3KWppn_b7P90:-499146009) ([raw file](https://github.com/sillsdev/icu-dotnet/blob/4cd1a8d8485f767eaff2c512e739d44e6d46b918/source/icu.net.tests/BreakIteratorTests.cs#L98)):*
<details><summary><i>Previously, ermshiperete (Eberhard Beilharz) wrote…</i></summary>
> These tests might be clearer to understand if you'd split them in two tests: something like `GetWordBoundaries_IncludeSpacesAndPunctuation` and `GetWordBoundaries_IgnoreSpacesAndPunctuation`.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/sillsdev/icu-dotnet/9)*
<!-- Sent from Reviewable.io -->

ermshiperete commented:
Reviewed 3 of 3 files at r2.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/sillsdev/icu-dotnet/9)*
<!-- Sent from Reviewable.io -->

ermshiperete commented:
@conniey Thanks!
